src/kaleidoscope/Kaleidoscope.h

```diff
@@ -42,6 +42,10 @@ extern HARDWARE_IMPLEMENTATION KeyboardHardware;
 
 typedef HARDWARE_IMPLEMENTATION::KeyAddr KeyAddr;
 
+inline constexpr KeyAddr allKeyAddr() {
```

Comment:
Having seen how we're using this in later commits, I think a name like allKeys might end up more readable and understandable than allKeyAddr

Reply:
I think `allKeyAddrs()` (or even `eachKeyAddr()`) would be better than `allKeys()`. When looking at this:
```c++
for (auto x : allKeys()) { ... }
```
…I would be expecting the `x` was of type `Key`, not `KeyAddr`.

More important, this is the best place for a comment describing how the iterator works, and how it's used. Someone skimming through the code and looking at this could be easily baffled, and someone who is confused by any of the range-based for loops that use it would first look here, not in `MatrixAddr.h` — that's two more searches away, and even then it's buried deep in the file without any good clues directing the reader to the relevant comment.

Reply:
…and I know I've said this before, but why use `auto` in those for loops and keep the type unclear, when you could remove all uncertainty and just use the actual type?

Reply:
I'd say that if we have KeyAddr in the thing we're iterating over, that's "Type spelled out nearby"

Reply:
allKeyAddrs() feels ok to me in a way that allKeyAddr() doesn't.

Reply:
Ok, let's go for `allKeyAddrs()`.

@gedankenexperimenter, there's a very informative article about the [AAA Style (Almost Always Auto)](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/), written by Herb Sutter.

Reply:
@noseglasses — I'll read it, but I'm very skeptical. Thanks for sharing the argument, in any case.

@obra — To some people, myself included, unless the actual type name `KeyAddr` appears somewhere nearby, that definitely is not "type spelled out nearby". If it were a class method (e.g. `KeyAddr::all()` or `KeyAddr::each()` or `KeyAddr::allKeyAddrs()`), then I would _almost_ agree with you, because `KeyAddr` at least appears as a type name in that case, rather than as part of a function name string. I would certainly guess correctly, but I would still find it impossible to just assume that it would do what I expect, and I with the code as it is, I would have a long odyssey to get to the comment in MatrixAddr.h where it is explained.

Reply:
You've argued your case. I'm happy with what Florian's got here. I'm resolving this conversation.